docs: update why not esbuild (#12228)
vitejs/vite: docs/guide/why.md

## Why Not Bundle with esbuild?

> When Vite was first created, `esbuild` was not used for builds as features like code-splitting and CSS handling were still a work in progress, so Rollup was used instead as it is more mature and flexible in these regards.
> 
> Today, `esbuild` is capable of handling these features, but Vite's adoption of Rollup's plugin API and architecture makes it hard to take advantage of `esbuild`'s performance, and even often incompatible. That said, we won't rule out the possibility of using `esbuild` for production builds if there are interests from the community.

Reviewer: I think we could word this in a more positive light. What about this angle?
```suggestion
`esbuild` has progressed a lot in the past years. It is now capable of handling features like code-splitting and CSS handling. Vite's adoption of Rollup's plugin API and architecture has been fundamental to its success. Framework authors have been able to innovate and explore new ideas, making Vite a shared base for their projects. We believe that Rollup offers a better performance-vs-flexibility tradeoff at this point.

That said, we won't rule out the possibility of using `esbuild` for production builds in the future. And we will keep taking advantage of new capabilities as they are released, as we have done with JS and CSS minification where `esbuild` allowed Vite to get a performance boost while avoiding disruption for its ecosystem.
```

Author: Good point. I like the second paragraph, but I'm not sure if the first answered the question of "Why not bundle with esbuild". It only mentioned that Rollup is better flexibility-wise over performance, that raises the question "Isn't esbuild flexible too?".

I think we can emphasize that because of Rollup's plugin API's widespread adoption that it would be a disruption if we break it, hence esbuild is less flexible. Maybe the first paragraph could be reworded as:

> `esbuild` has progressed a lot in the past years, and is now capable of handling features like code-splitting and CSS. As Vite adopted Rollup's plugin API and infrastructure, which contributed to its success in the ecosystem, it conflicts with `esbuild`'s plugin API and would be a fundamental breaking change. For the time being, we believe that Rollup offers a better performance-vs-flexibility tradeoff.

Reviewer: I think the two lines are a bit contradictory. If we believe that Rollup offers a better performance-vs-flexibility tradeoff, then even if we would still stay in Rollup even if there was a way to magically move everyone to esbuild, no?

Maybe something like:

> `esbuild` has progressed a lot in the past years, and it can now handle features like code-splitting and CSS. `esbuild`'s plugin API is now flexible enough to be used for production bundling. However, the adoption of Rollup's plugin API and infrastructure heavily contributed to Vite success in the ecosystem, and both APIs aren't compatible. For the time being, we believe that Rollup offers a better performance-vs-flexibility tradeoff.

Reviewer 2: > `esbuild` has progressed a lot in the past years. It is now capable of handling features like code-splitting and CSS handling.

I might either remove these sentences or use something more like @bluwy's original version of them. I think this version is missing the historical context and so it's unclear how they relate to the question.

Another reason we could mention: esbuild is written in Go, but much of the JS ecosystem is written in JS. We wouldn't get a lot of the speed benefits because we need to call out to these JS libraries such as the Svelte compiler

Reviewer: I'm fine with either, your last paragraph is what I'm aiming at myself. I just would like to avoid sounding like if we would start over now, we would go with esbuild. I still think that rollup is a better fit for this stage of the ecosystem, we keep using it on its own merits and not only because it is incompatible with esbuild API.

Author: Honestly I thought if we were to start over, we would use esbuild's API 😄 Though it _is_ less flexible than Rollup when configuring advanced stuff.

I don't think saying "the JS ecosystem is slowing down builds even if written in Go" is true, esbuild is technically faster in bundling still where Rollup is slower when leaving out plugins.

I'll take a stab at updating this again with these discussion and we can tweak further again.

> That said, `esbuild` has progressed a lot in the past years, and we won't rule out the possibility of using `esbuild` for production builds in the future. We will keep taking advantage of new capabilities as they are released, as we have done with JS and CSS minification where `esbuild` allowed Vite to get a performance boost while avoiding disruption for its ecosystem.

Author: I removed the history bit of `esbuild` as it's not the main reason to the question anymore. I hope the first paragraph now directly reflects why we didn't go with esbuild.

## Why Not Bundle with esbuild?

> Vite's adoption of Rollup's plugin API and infrastructure heavily contributed to its success in the ecosystem, however, it isn't compatible with `esbuild`'s API. For the time being, we believe that Rollup offers a better performance-vs-flexibility tradeoff.

Reviewer: One last suggestion only reordering the lines 👀
I'm fine with either though
```suggestion
Vite's current plugin API isn't compatible with using `esbuild` as a bundler. For the time being, we believe that Rollup offers a better performance-vs-flexibility tradeoff. Vite's adoption of Rollup's plugin API and infrastructure heavily contributed to its success in the ecosystem.
```

Reviewer 2: It might be clearer to just say "better flexibility" than "better performance-vs-flexibility tradeoff" as the latter requires some reader knowledge of which one is faster and which one is more flexible to be able to interpret

Reviewer: Is this more clear?

> Vite's current plugin API isn't compatible with using `esbuild` as a bundler. In spite of `esbuild` being faster, Vite's adoption of Rollup's flexible plugin API and infrastructure heavily contributed to its success in the ecosystem. For the time being, we believe that Rollup offers a better performance-vs-flexibility tradeoff.

Author: I like that too. Will change to that 👍 
```suggestion
Vite's current plugin API isn't compatible with using `esbuild` as a bundler. In spite of `esbuild` being faster, Vite's adoption of Rollup's flexible plugin API and infrastructure heavily contributed to its success in the ecosystem. For the time being, we believe that Rollup offers a better performance-vs-flexibility tradeoff.
```
